pkg/neotest/coverage.go

```diff
@@ -0,0 +1,205 @@
+package neotest
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"os"
+	"strconv"
+	"sync"
+	"testing"
+
+	"github.com/nspcc-dev/neo-go/pkg/compiler"
+	"github.com/nspcc-dev/neo-go/pkg/util"
+	"github.com/nspcc-dev/neo-go/pkg/vm"
+	"github.com/nspcc-dev/neo-go/pkg/vm/opcode"
+)
+
+const (
+	goCoverProfileFlag  = "test.coverprofile"
+	// disableNeotestCover is name of the environmental variable used to explicitly disable neotest coverage.
+	disableNeotestCover = "DISABLE_NEOTEST_COVER"
+)
+
+var (
+	// coverageLock protects all vars below from concurrent modification when tests are run in parallel.
+	coverageLock sync.Mutex
+	// rawCoverage maps script hash to coverage data, collected during testing.
+	rawCoverage = make(map[util.Uint160]*scriptRawCoverage)
+	// flagChecked is true if `go test` coverage flag was checked at any point.
+	flagChecked bool
+	// coverageEnabled is true if coverage is being collected on test execution.
+	coverageEnabled bool
+	// coverProfile specifies the file all coverage data is written to, unless empty.
+	coverProfile = ""
+)
+
+type scriptRawCoverage struct {
+	debugInfo      *compiler.DebugInfo
+	offsetsVisited []int
+}
+
+type coverBlock struct {
+	// Line number for block start.
+	startLine uint
+	// Column number for block start.
+	startCol uint
+	// Line number for block end.
+	endLine uint
+	// Column number for block end.
+	endCol uint
+	// Number of statements included in this block.
+	stmts uint
+	// Number of times this block was executed.
+	counts uint
+}
+
+// documentName makes it clear when a `string` maps path to the script file.
+type documentName = string
+
+func isCoverageEnabled() bool {
+	coverageLock.Lock()
+	defer coverageLock.Unlock()
+
+	if flagChecked {
+		return coverageEnabled
+	}
+	defer func() { flagChecked = true }()
+
+	if isDisabledStr, ok := os.LookupEnv(disableNeotestCover); ok {
+		isDisabled, err := strconv.ParseBool(isDisabledStr)
+		if err != nil {
+			panic(fmt.Sprintf("coverage: error when parsing environment variable '%s', expected bool, but got '%s'", disableNeotestCover, isDisabledStr))
+		}
+		coverageEnabled = !isDisabled
+		return coverageEnabled
+	}
```

Comment:
I think we need to combine this env with go coverage flag. I.e. if this env is set to false but coverage flag is missing, then we can't enable coverage. The source of truth still should be a `goCoverProfileFlag`.